packages/core/typing_test.tsx

```diff
@@ -0,0 +1,296 @@
+      <ul className="list-group">
+        {errors.map((error, i) => {
+          return (
+            <li key={i} className="list-group-item text-danger">
+              {error.stack}
+            </li>
+          );
+        })}
+      </ul>
+    </div>
+  );
+}
+
+export class Example extends React.Component<any, IExampleState> {
+  public state: IExampleState = {
+    formData: {
+      firstName: 'Chuck',
+      lastName: 'Norris',
+      age: 75,
+      bio: 'Roundhouse kicking asses since 1940',
+      password: 'noneed',
+    },
+  };
+
+  constructor(props: any) {
+    super(props);
+  }
+
+  public render() {
+    return (
+      <div className="react-jsonschema-form-example">
+        {
+          <Form
+            schema={schema}
+            uiSchema={uiSchema}
+            showErrorList={false}
+            noValidate={false}
+            noHtml5Validate={false}
+            formData={this.state}
+            ErrorList={ErrorListExample}
+            onChange={formData => this.setState({ formData })}
+            customFormats={{
+              'phone-us': /\(?\d{3}\)?[\s-]?\d{3}[\s-]?\d{4}$/,
+            }}
+          />
+        }
+      </div>
+    );
+  }
+}
+
+export class ExampleSchemaField extends React.Component<SchemaFieldProps> {
+  constructor(props: SchemaFieldProps) {
+    super(props);
+  }
+
+  public render() {
+    return <SchemaField {...this.props} />;
+  }
+}
+
+interface FuncExampleProps {
+  formData: object;
+  onError: (e: ErrorSchema) => void;
+  onChange: (e: any) => void;
+}
+
+export const FuncExample = (props: FuncExampleProps) => {
+  const { formData, onChange, onError } = props;
+  return (
+    <Form
+      schema={schema}
+      uiSchema={uiSchema}
+      showErrorList={false}
+      noValidate={false}
+      noHtml5Validate={false}
+      formData={formData}
+      ErrorList={ErrorListExample}
+      onChange={(formData, errorSchema) => {
+        onChange(formData);
+        errorSchema && onError(errorSchema);
+      }}
+    />
+  );
+};
+
+export const BooleanCustomWidget: React.SFC<WidgetProps> = props => (
+  <input onFocus={() => props.onFocus('id', true)} onBlur={() => props.onFocus('id', true)} />
+);
+
+export const NumberCustomWidget: React.SFC<WidgetProps> = props => (
+  <input onFocus={() => props.onFocus('id', 0)} onBlur={() => props.onFocus('id', 0)} />
+);
+
+export const StringCustomWidget: React.SFC<WidgetProps> = props => (
+  <input onFocus={() => props.onFocus('id', 'value')} onBlur={() => props.onFocus('id', 'value')} />
+);
+
+export const NullCustomWidget: React.SFC<WidgetProps> = props => (
+  <input onFocus={() => props.onFocus('id', null)} onBlur={() => props.onFocus('id', null)} />
+);
+
+export const withThemeExample = () => {
+  const Form = withTheme({
+    showErrorList: false,
+    noValidate: false,
+    noHtml5Validate: false,
+  });
+  const forwardedRef = React.useRef<Form<any>>(null);
+
+  return <Form schema={schema} ref={forwardedRef} />;
+};
```

Comment:
@zepatrik  this was what I was getting at in the other comment about the return types for withTheme

Reply:
So you want `withTheme` to be generic and not provide the generic when creating the `Form` component? That makes sense, but I guess as you said it is not really possible. My workaround does not fully fix this but I guess in most cases you will have the generic type of the final form anyways when you call the factory `withTheme`. Having this non-ideal typing is at least better than having none at all in my opinion.

Reply:
@zepatrik, no the generic to withTheme doesn't allow (and shouldn't) formData, so the formdata passed to the resulting form from withTheme is always any. 

withTheme's T generic argument is just being inferred from the ThemeProps<T>, which would in this example be 

```
{
    showErrorList: false,
    noValidate: false,
    noHtml5Validate: false,
  }
  ```
  
  which is correct
  
  returning: `typeof Form` is the same as returning a Form class component (from typescripts knowlege), and _does_ correctly allow my example above to work.  It doesn't mirror the code here as directly, but it is properly to the spirit of this component factory.
  
  it could also, similarly just return a `<U>(props: FormProps<U> & React.RefAttributes<Form<U>>>) => JSX.Element`, however this seems less clear than just returning a Form from withTheme (making it more correctly appear like a factory)

Reply:
Ok I'm sorry, I am not really deep into this topic right now as the PR is already half a year old :sweat_smile: 
Do you want to take this over or fix it in a new one? I think it is a good idea to have this type test added as I did in this PR.
Feel free to open a new PR based on these commits.

Reply:
haha, no worries!  I opened #2279 the other day, without seeing this first (I probably should have searched!).   np, i'll pull some of these changes into that branch or those into this one and consolidate everything when I get a minute, thanks!